Readd Meteor Filler

koolkrafter5 opened the pull request:
Reimplement the concept of "filler" blocks to Blood Magic meteors.
Look for filler block list and fillerChance (chance for any individual block to be replaced with filler out of 100) in meteor config files in config/BloodMagic/meteors. Default filler block is minecraft:stone and default fillerChance is 0.
NEI meteor pages correctly show how much filler is expected in a meteor.
Fix an error in the block yield estimation function for more accurate results in NEI.

koolkrafter5 commented:
Here is an example config file for a meteor that uses filler. Meteor is summoned with a 4:1 ratio of coal ore to iron ore, and 20% is replaced with sandstone as a filler block. Tenebrae, crystallos, and incendium may be added to replace the sandstone. Terrae/orbis terrae currently do not increase the filler chance.

[CoalBlockMeteor.json](https://github.com/user-attachments/files/17783257/CoalBlockMeteor.json)

AbdielKavash commented on src/main/java/WayofTime/alchemicalWizardry/client/nei/NEIMeteorRecipeHandler.java:
```diff
@@ -50,7 +54,7 @@ public CachedMeteorRecipe(MeteorParadigm meteor, ItemStack focusStack) {
 
                 List<String> tooltips = new ArrayList<>();
-                float chance = component.getChance() / totalMeteorWeight;
+                float chance = component.getChance() / totalComponentWeight * (float) (1 - fillerChance / 100.0);
```
I would precompute:
```java
float fillerRatio = fillerChance / 100.0;
float componentRatio = 1 - fillerRatio;
```
outside of the loop, and then multiply chances by this.

koolkrafter5 replied:
Addressed in [b5f55eb](https://github.com/GTNewHorizons/BloodMagic/pull/60/commits/b5f55eb79eadb8e225dff4fafff22f3db3a52a2f)

AbdielKavash commented on src/main/java/WayofTime/alchemicalWizardry/client/nei/NEIMeteorRecipeHandler.java:
```diff
@@ -40,7 +42,9 @@ public CachedMeteorRecipe(MeteorParadigm meteor, ItemStack focusStack) {
 
-            float totalMeteorWeight = meteor.getTotalMeteorWeight();
+            float totalComponentWeight = meteor.getTotalComponentWeight();
+            float totalFillerWeight = meteor.getTotalFillerWeight();
```
Very small nitpick, either move `totalFillerWeight` down to where you get `sortedFiller` (i.e., initialize all variables where they are needed), or move `sortedFiller` here (i.e., initialize everything at the start).

koolkrafter5 replied:
Addressed in [49e0274](https://github.com/GTNewHorizons/BloodMagic/pull/60/commits/49e0274d9ea18e2c9db6679ef36285538417cc92)

AbdielKavash commented on src/main/java/WayofTime/alchemicalWizardry/client/nei/NEIMeteorRecipeHandler.java:
```diff
@@ -61,10 +65,50 @@ public CachedMeteorRecipe(MeteorParadigm meteor, ItemStack focusStack) {
                 }
             }
+
+            List<MeteorParadigmComponent> sortedFiller = new ArrayList<>(meteor.fillerList);
```
Move this inside of the conditional below. (Or to the top of the method, as I mentioned in an earlier comment.)

koolkrafter5 replied:
Addressed in [49e0274](https://github.com/GTNewHorizons/BloodMagic/pull/60/commits/49e0274d9ea18e2c9db6679ef36285538417cc92)

AbdielKavash commented on src/main/java/WayofTime/alchemicalWizardry/common/summoning/meteor/MeteorParadigm.java:
```diff
@@ -21,29 +21,41 @@
+        new MeteorParadigm(focusStack, radius, cost, 0);
+    }
+
+    public MeteorParadigm(ItemStack focusStack, int radius, int cost, int fillerWeight) {
```
Name the last argument `fillerChance`, like the field.

koolkrafter5 replied:
Addressed in [9646ade](https://github.com/GTNewHorizons/BloodMagic/pull/60/commits/9646ade74859609e020ca76b04f64ac322831c31)

AbdielKavash commented on src/main/java/WayofTime/alchemicalWizardry/client/nei/NEIMeteorRecipeHandler.java:
```diff
@@ -61,10 +65,50 @@ public CachedMeteorRecipe(MeteorParadigm meteor, ItemStack focusStack) {
+                    int yPos = 37 + 18 * row;
+
+                    List<String> tooltips = new ArrayList<>();
+                    float chance = filler.getChance() / totalFillerWeight * (float) (fillerChance / 100.0);
```
Precompute `fillerRatio` outside of the loop, as mentioned above.

koolkrafter5 replied:
Addressed in [b5f55eb](https://github.com/GTNewHorizons/BloodMagic/pull/60/commits/b5f55eb79eadb8e225dff4fafff22f3db3a52a2f)

AbdielKavash commented on src/main/java/WayofTime/alchemicalWizardry/common/summoning/meteor/Meteor.java:
```diff
@@ -21,6 +21,8 @@ public class Meteor {
     public String focusModId;
     public String focusName;
     public int focusMeta;
+    private String[] filler;
+    private int fillerChance;
```
Either make both of these `public` too, or (better) make anything that is not needed outside of the class also `private`.

AbdielKavash commented on src/main/java/WayofTime/alchemicalWizardry/common/summoning/meteor/MeteorParadigm.java:
```diff
@@ -143,43 +164,9 @@ public void createMeteorImpact(World world, int x, int y, int z, boolean[] flags
-                    }
-
-                    if (!hasPlacedBlock) {
+                    if (world.rand.nextInt(100) >= fillerChance) {
```
`if (fillerChance <= 0 || world.rand.nextInt(100) >= fillerChance)` for short-circuiting

AbdielKavash reviewed (changes requested):
Extra comments:

- [ ] Add filler blocks to `NEIMeteorRecipeHandler.loadCraftingRecipes`, this means the meteors will show up when the player searches for recipes for e.g. stone. I can't see this ever being particularly *useful*, but it is consistent.
- [ ] Not a huge fan of how the replacement filler (ice/obsidian/etc.) is currently generated in `MeteorParadigm` starting from around line 170. I know this is not your code. But since you now have a generic `setMeteorBlock`, you could precompute a new `List<MeteorParadigmComponent>` at the place where you check for reagents, and then just call `setMeteorBlock` on this list instead of `fillerList` if any reagents are present. This also future-proofs it for if/when we want to add more reagents.
- [ ] Something to consider, and honestly I am not sure what is the best solution or even if this is a problem at all: Currently you have `stone` hardcoded as a default filler block, both in the NEI handler, and in the meteor spawning code.
It would be perhaps more sensible to *always* initialize `MeteorParadigm.fillerList`; either in the constructor or when you are reading filler from config -- if `fillerChance` is not 0, and `fillerList` is empty, then add a stone block.
Alternatively, `MeteorParadigmComponent` could have a `static` field that defines what the default block is?
Since it is literally just stone I don't consider this a big "magic value" issue; but still if for whatever reason someone wants to change the default filler block, it is better to have it declared in one place.

koolkrafter5 commented:
> Extra comments:
> 
>     * [*]  Add filler blocks to `NEIMeteorRecipeHandler.loadCraftingRecipes`, this means the meteors will show up when the player searches for recipes for e.g. stone. I can't see this ever being particularly _useful_, but it is consistent.

Addressed in [d9d70bb](https://github.com/GTNewHorizons/BloodMagic/pull/60/commits/d9d70bbd6b1fbcffb5d5573bdcf5ea311f810ad3).

>     * [*]  Not a huge fan of how the replacement filler (ice/obsidian/etc.) is currently generated in `MeteorParadigm` starting from around line 170. I know this is not your code. But since you now have a generic `setMeteorBlock`, you could precompute a new `List<MeteorParadigmComponent>` at the place where you check for reagents, and then just call `setMeteorBlock` on this list instead of `fillerList` if any reagents are present. (Or even better, replace `sortedFiller` by this list, so that it "magically" happens on its own.) This also future-proofs it for if/when we want to add more reagents.

Addressed in [1af2bbb](https://github.com/GTNewHorizons/BloodMagic/pull/60/commits/1af2bbb40f2b6812897a3746aeaf04dddc407337) by making a new local fillerList and replacing it if any reagents are present.

>     * [*]  Something to consider, and honestly I am not sure what is the best solution or even if this is a problem at all: Currently you have `stone` hardcoded as a default filler block, both in the NEI handler, and in the meteor spawning code.
>       It would be perhaps more sensible to _always_ initialize `MeteorParadigm.fillerList`; either in the constructor or when you are reading filler from config -- if `fillerChance` is not 0, and `fillerList` is empty, then add a stone block.
>       Alternatively, `MeteorParadigmComponent` could have a `static` field that defines what the default block is?
>       Since it is literally just stone I don't consider this a big "magic value" issue; but still if for whatever reason someone wants to change the default filler block, it is better to have it declared in one place.

Addressed in [3fe4b5b](https://github.com/GTNewHorizons/BloodMagic/pull/60/commits/3fe4b5b588aebe5155d84f1cf1179b67813eddae) by setting the fillerList to stone with a weight of 1 when the weight is 0 or there is no block list supplied.

Minecraft crashes in MeteorParadigm.parseStringArray() if a filler or block list in a meteor config is written improperly. It appears to be a problem with parseInt in the legacy config section, and removing that section prevents that crash. Should there be more error checking there, or will that be removed soon enough that it won't be a problem?

It also crashes from a MalformedJsonException in Meteor.loadConfig() if a meteor config is malformed (E.g. missing a comma). Should I move that to a try catch block and write an error to the log if that happens so Minecraft can continue to load?

AbdielKavash commented on src/main/java/WayofTime/alchemicalWizardry/client/nei/NEIMeteorRecipeHandler.java:
```diff
@@ -61,10 +66,44 @@ public CachedMeteorRecipe(MeteorParadigm meteor, ItemStack focusStack) {
+                    row++;
+                }
+
+                List<MeteorParadigmComponent> sortedFiller = new ArrayList<>(meteor.fillerList);
```
This is not actually sorted.

koolkrafter5 replied:
Addressed by [9c0be69](https://github.com/GTNewHorizons/BloodMagic/pull/60/commits/9c0be69a1719c5f63d1fdb3c105e7476080c8533)

AbdielKavash commented on src/main/java/WayofTime/alchemicalWizardry/common/summoning/meteor/MeteorParadigm.java:
```diff
@@ -115,22 +127,38 @@ public void createMeteorImpact(World world, int x, int y, int z, boolean[] flags
+        int fillerChance = this.fillerChance;
         if (hasOrbisTerrae) {
             newRadius += 2;
+            fillerChance *= 1.12;
         } else if (hasTerrae) {
             newRadius += 1;
+            fillerChance *= 1.6;
```
- [ ] Are you sure these two values are not swapped? Here Terrae increases filler ratio significantly more than Orbis Terrae; which contradicts your comments and FTB wiki on the expected behavior.
- [ ] This doesn't break anything (I think), but I would still sanitize this by clamping `fillerChance` to a maximum of 100.

koolkrafter5 replied:
> [x]  Are you sure these two values are not swapped? Here Terrae increases filler ratio significantly more than Orbis Terrae; which contradicts your comments and FTB wiki on the expected behavior. [Edit: you wrote Terrae should be 1.06, but you have 1.6 in code - that is probably the error.]

Should have been 1.06, addressed by [c8aea92](https://github.com/GTNewHorizons/BloodMagic/pull/60/commits/c8aea9243f9f4e42143b1770d4b425589e1c3e87).

> [x]  This doesn't break anything (I think), but I would still sanitize this by clamping `fillerChance` to a maximum of 100.

Addressed by [fa6cd74](https://github.com/GTNewHorizons/BloodMagic/pull/60/commits/fa6cd740536b0c0dccfdc84dd180e2f2d6df6437).

AbdielKavash commented on src/main/java/WayofTime/alchemicalWizardry/common/summoning/meteor/MeteorRegistry.java:
```diff
@@ -15,10 +16,20 @@ public static void registerMeteorParadigm(MeteorParadigm paradigm) {
+            if (fillerChance > 0 && fillerList != null && fillerList.length > 0) {
+                meteor.parseStringArray(fillerList, true);
+            } else {
+                meteor.fillerList.add(new MeteorParadigmComponent(new ItemStack(Blocks.stone), 1));
```
I am wondering what the right conditions for running this actually should be.

* If `fillerChance` is 0 and `fillerList` is empty (e.g., all current GTNH meteors), you still add a stone block as filler. This will have no observable impact as any place filler is used specifically checks for `fillerChance` being non-zero.
* If `fillerChance` is 0 and `fillerList` is *not* empty, you still overwrite whatever the filler list was in the config with stone. I don't know what the expected behavior should be in this case; but probably not this. (Hypothetically if there was a reagent that increases filler chance by a constant value, rather than multiplicatively, this would lead to the filler defined in config being ignored.)

In my opinion this should be something like:

* `fillerChance > 0`, `fillerList` not empty: Parse `fillerList`.
* `fillerChance > 0`, `fillerList` empty: Add stone.
* `fillerChance <= 0`, `fillerList` not empty: Parse `fillerList`, even though it will not be used.
* `fillerChance <= 0`, `fillerList` empty: Initialize list to some default empty value (null, empty list, stone?)

koolkrafter5 replied:
Addressed by [2f96807](https://github.com/GTNewHorizons/BloodMagic/pull/60/commits/2f968072b5053645c33aad79faa00331f991f9fb) by reading the filler list regardless of the value of fillerChance. Setting fillerList to a null/empty list would cause problems with a future reagent that adds a fixed value to fillerChance, so it will just be initialized to stone.

The filler list is not shown in NEI if the filler chance is 0. That information is unused with the current implementation, but it may be useful in the future with a reagent that adds to the filler chance. Should I do something to address that? You can mark this as resolved if not.

AbdielKavash replied:
> The filler list is not shown in NEI if the filler chance is 0. That information is unused with the current implementation, but it may be useful in the future with a reagent that adds to the filler chance. Should I do something to address that? You can mark this as resolved if not.

Looks good; honestly I'm not sure what the "most correct" behavior here would be. If we (or anyone else) decide to give it some meaningful interpretation later, we can always go back and change it.

AbdielKavash reviewed (changes requested):
Sorry about the delay; functionally everything looks good, check my comments for last couple of technical details.

AbdielKavash reviewed (approved):
All looks good now!

As I said above we will wait until 2.7 to actually merge this, so if you have any other ideas feel free to continue working on it.

koolkrafter5 commented:
> All looks good now!
> 
> As I said above we will wait until 2.7 to actually merge this, so if you have any other ideas feel free to continue working on it.

Awesome! I think I'll look into making the reagent effects for meteors configurable next.

AbdielKavash commented:
> > All looks good now!
> > As I said above we will wait until 2.7 to actually merge this, so if you have any other ideas feel free to continue working on it.
> 
> Awesome! I think I'll look into making the reagent effects for meteors configurable next.

We should probably have a discussion about this before any big changes, because it might significantly affect balance. (Which might not be a *bad* thing necessarily.)

One thing I would like though is to keep the effect of the "base" reagents (as on the [FTB wiki](https://ftbwiki.org/Mark_of_the_Falling_Tower)) unchanged, so as to not confuse people googling information about the rituals. There are plenty of other reagents we can use to add fun effects without having to change those.

koolkrafter5 commented:
> One thing I would like though is to keep the effect of the "base" reagents (as on the [FTB wiki](https://ftbwiki.org/Mark_of_the_Falling_Tower) for example) unchanged...

That was my plan for when I would add this. The current reagents would stay the same in the default configuration, but it would leave the ability for people to tweak their effects and for other reagents to be added.

Latest commit was for something that simplifies the code added here *and* is useful for the reagent stuff, but I will make another PR from a different branch of my BM repo for the rest of the reagent configuration feature.

Could we possibly add dynamic pages to NEI that explain the reagent effects? I feel like that would make these and any future changes more discoverable for players.

AbdielKavash commented:
> Could we possibly add dynamic pages to NEI that explain the reagent effects? I feel like that would make these and any future changes more discoverable for players.

That would be awesome, though ideally it should be done in a way that it works for all rituals.

Dream-Master commented:
Is this frozen or should be added for 2.7.0 ?

AbdielKavash commented:
This currently does not have any gameplay impact, as no meteors that are currently defined contain any filler blocks. Thus I think it is safer to delay until after 2.7. Then we can adjust the composition of various meteors, and possibly add filler where needed or reasonable. (This would be a balance change.)
